processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -147,23 +135,14 @@ protected Integer addToFacts(
             makeColumnSelectorFactory(agg, rowSupplier, deserializeComplexMetrics)
         );
       }
-      final Integer rowIndex = indexIncrement.getAndIncrement();
-
-      concurrentSet(rowIndex, aggs);
-
-      // Last ditch sanity checks
-      if (numEntries.get() >= maxRowCount && !facts.containsKey(key)) {
+      if (numEntries.incrementAndGet() > maxRowCount) {
+        numEntries.decrementAndGet();
         throw new IndexSizeExceededException("Maximum number of rows [%d] reached", maxRowCount);
       }
-      final Integer prev = facts.putIfAbsent(key, rowIndex);
-      if (null == prev) {
-        numEntries.incrementAndGet();
```

Comment:
Is there a reason we are removing a lot of the existing concurrency checks? They were there to prevent nasty race conditions we've experienced in the past?